From 14eb019687639211cf2185d5d2a178a05419f86b Mon Sep 17 00:00:00 2001
From: Chin-Ting Kuo <chin-ting_kuo@aspeedtech.com>
Date: Sun, 8 Feb 2026 13:45:20 +0800
Subject: [PATCH] mtd: spi: Add support for Winbond W25QxxRV serial NOR flash
 devices

On W25QxxRV parts, the configuration register cannot be programmed
using the 01h (status register program) command. Use the 31h
command instead. If the 01h command is used, the last 8 clock
cycles are ignored.

Signed-off-by: Chin-Ting Kuo <chin-ting_kuo@aspeedtech.com>
Change-Id: I17dcb3b7a20fee03952c06a3be69105950107b88
(cherry picked from commit 8b4033aa36f4db0444606b31c63538641c285099)
Signed-off-by: Subraja M <subrajam@ami.com>
---
 drivers/mtd/spi/spi-nor-core.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/mtd/spi/spi-nor-core.c b/drivers/mtd/spi/spi-nor-core.c
index dfa191845a..f0a2bea415 100644
--- a/drivers/mtd/spi/spi-nor-core.c
+++ b/drivers/mtd/spi/spi-nor-core.c
@@ -1343,6 +1343,27 @@ static int write_sr_cr(struct spi_nor *nor, u8 *sr_cr)
 		return ret;
 	}
 
+#if defined(CONFIG_SPI_FLASH_WINBOND)
+	if (JEDEC_MFR(nor->info) == SNOR_MFR_WINBOND) {
+		write_enable(nor);
+
+		ret = nor->write_reg(nor, SPINOR_OP_WINBOND_WRSR2, &sr_cr[1], 1);
+		if (ret < 0) {
+			dev_dbg(nor->dev,
+				"error while writing configuration register\n");
+			return -EINVAL;
+		}
+
+		ret = spi_nor_wait_till_ready(nor);
+		if (ret) {
+			dev_dbg(nor->dev,
+				"timeout while writing configuration register\n");
+			return ret;
+		}
+
+		write_disable(nor);
+	}
+#endif
 	return 0;
 }
 
-- 
2.53.0

